packages/core/src/connectors/walletConnect.ts

```diff
@@ -56,6 +73,58 @@ export class WalletConnectConnector extends Connector<
       provider.on('chainChanged', this.onChainChanged)
       provider.on('disconnect', this.onDisconnect)
 
+      if (this.version === '2') {
+        provider.on('session_delete', this.onDisconnect)
+        provider.on('display_uri', this.onDisplayUri)
+
+        await Promise.race([
+          provider.connect({
+            namespaces: {
+              [sharedConfig.namespace]: {
+                methods: [
+                  'eth_sendTransaction',
+                  'eth_sign',
+                  'eth_signTransaction',
+                  'eth_signTypedData',
+                  'personal_sign',
+                ],
+                events: ['accountsChanged', 'chainChanged'],
+                chains: this.chains.map(
+                  (chain) => `${sharedConfig.namespace}:${chain.id}`,
+                ),
+                rpcMap: this.chains.reduce(
+                  (rpc, chain) => ({
+                    ...rpc,
+                    [chain.id]: chain.rpcUrls.default,
+                  }),
+                  {},
+                ),
+              },
+            },
+          }),
+          ...(this.options.qrcode
+            ? // When using WalletConnect QR Code Modal, open modal and listen for close callback.
+              // If modal is closed, reject promise so `catch` block for `connect` is called.
+              [
+                new Promise((_res, reject) =>
+                  provider.on('display_uri', async (uri: string) => {
+                    const { default: Modal } = await import(
+                      '@walletconnect/qrcode-modal'
+                    )
+                    Modal.open(uri, () => reject(new Error('user rejected')))
+                  }),
+                ),
+              ]
+            : []),
+        ])
+
+        // If execution reaches here, connection was successful and we can close modal.
+        if (this.options.qrcode) {
+          const { default: Modal } = await import('@walletconnect/qrcode-modal')
+          Modal.close()
+        }
```

Comment:
This assumption feels a little shaky. Is this the best way to know if `provider.connect` succeeded?

Reply:
I think it should be fine. If the connection request is declined by the user an exception is thrown and is being captured correctly `catch` block